framework/infrastructure/gcp/cloud_run.py

```diff
@@ -0,0 +1,131 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import logging
+
+from googleapiclient import discovery
+
+from framework.infrastructure import gcp
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TEST_PORT = 8080
+DEFAULT_TIMEOUT = 600
+
+
+class CloudRunApiManager(
+    gcp.api.GcpStandardCloudApiResource, metaclass=abc.ABCMeta
+):
+    project: str
+    region: str
+    _parent: str
+    service: discovery.Resource
+    api_manager: gcp.api.GcpApiManager
+
+    def __init__(self, project: str, region: str):
+        if not project:
+            raise ValueError("Project ID cannot be empty or None.")
+        if not region:
+            raise ValueError("Region cannot be empty or None.")
+        self.api_manager = gcp.api.GcpApiManager(
+            v2_discovery_uri="https://run.googleapis.com/$discovery/rest?"
+        )
+        self.project = project
+        self.region = region
+        service: discovery.Resource = self.api_manager.cloudrun("v2")
+        self.service = service
+        self._parent = f"projects/{self.project}/locations/{self.region}"
+        super().__init__(self.service, project)
+
+    @property
+    def api_name(self) -> str:
+        """Returns the API name for Cloud Run."""
+        return "run"
+
+    @property
+    def api_version(self) -> str:
+        """Returns the API version for Cloud Run."""
+        return "v2"
+
+    def create_cloud_run_resource(
+        self, service: discovery.Resource, service_name: str, body: dict
+    ):
+        service.projects().locations().services().create(
+            parent=self._parent, serviceId=service_name, body=body
+        ).execute()
+
+    def get_cloud_run_resource(
+        self, service: discovery.Resource, service_name: str
+    ):
+        return self._get_resource(
+            collection=service.projects().locations().services(),
+            full_name=self.resource_full_name(
+                service_name, "services", self.region
+            ),
+        )
+
+    def get_service_uri(self, service_name: str) -> str:
+        response = self.get_cloud_run_resource(self.service, service_name)
+        return response.get("uri")
+
+    def delete_cloud_run_resource(
+        self, service: discovery.Resource, service_name: str
+    ):
+        service.projects().locations().services().delete(
+            name=self.resource_full_name(service_name, "services", self.region)
+        ).execute()
+
+    def deploy_service(
```

Comment:
This a business logic not tied to wrapping API calls. It should be in an ApiManager, which controls "what" resources we create, while this library controls "how" we talk to the API itself. Move this method to `TrafficDirectorCloudRunManager`.

Reply:
The idea is that `TrafficDirectorAppNetManager` will eventually be renamed to `AppNetMeshManager` or `AppNetMeshManager`, as well as the rest of the classes. `TrafficDirectorManager` will be `TrafficDirectorMeshManager` too.

All these are resource managers holding the state of the mesh under test - among with any test-specific resources we need.